stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
+    // the trace.
+    let maxReturnValueCount = outputs.count
+    debugLog("""
+               Executing trace func \(tracedFunctionName) with up to \
+               \(maxReturnValueCount) return values.
+               """)
+    var returnValues: [CTensorHandle?]
+    returnValues = Array(repeating: nil, count: maxReturnValueCount)
+    var outputReturnValueCount = Int32(maxReturnValueCount)
+    TFE_Execute(op, &returnValues, &outputReturnValueCount, status) 
+    checkOk(status)
+    debugLog("""
+               returnValues.count=\(returnValues.count), \
+               outputReturnValueCount=\(outputReturnValueCount).
+               """)
+    internalConsistencyCheck(outputReturnValueCount <= returnValues.count)
+
+    // Now that all the output elements have been filled in, remove a level of
+    // optional, and also add concrete outputs.
+    var traceGraphOutputs: [CTensorHandle] = []
+    // Points to an element in `returnValues`.
+    var returnValueIdx = 0
+    for output in outputs {
+      internalConsistencyCheck(TFE_TensorHandleIsConcrete(output) == 0)
+      internalConsistencyCheck(returnValues[returnValueIdx] != nil)
+      traceGraphOutputs.append(returnValues[returnValueIdx]!)
+      returnValueIdx += 1
+    }
+    internalConsistencyCheck(returnValueIdx == outputReturnValueCount)
+    return traceGraphOutputs
+  }
+
+  // Finalize the trace graph function.
+  private func finalize(tracedFunctionName: String,
+                        outputs: [CTensorHandle]) -> CTFFunction {
+    var symbolicOutputs: [TF_Output] = []
+    for (i, output) in outputs.enumerated() {
+      debugLog("Adding symbolic output \(i) as a trace graph func output.")
+      symbolicOutputs.append(TFE_GetTFOutputFromTensorHandle(output ,status))
+      checkOk(status)
+    }
+
+    debugLog("Trace graph fn has \(symbolicOutputs.count) outputs.")
+    let theTFFunction =
+      TF_GraphToFunction(graph, tracedFunctionName,
+                         /*append_hash_to_fn_name*/ 0,
+                         /*num_opers*/ -1,
+                         /*opers*/ nil,
+                         /*numinputs*/ Int32(symbolicInputs.count),
+                         /*inputs*/ symbolicInputs,
+                         /*noutputs*/ Int32(symbolicOutputs.count),
+                         /*outputs*/ symbolicOutputs,
+                         /*outputnames*/ nil,
+                         /*functionoptions*/ nil, "", status)
+    checkOk(status)
+
+    if _RuntimeConfig.printsDebugLog {
+      var len: Int = 0
+      let funcDebugStr = TF_FunctionDebugString(theTFFunction, &len)!
+      debugLog("The traced function is:\n\(String(cString: funcDebugStr))")
+      free(funcDebugStr)
+    }
+
+    return theTFFunction!
+  }
+}
+
+// This enum keeps track of whether we are building or executing a trace.
+private enum TracingState {
+  case notTracing
+  case tracing(TraceContext)
+
+  // Return nil if we are not in tracing mode.
+  var getFunction : TraceContext? {
```

Comment:
* Avoid using `get` with an object in Swift names.
* Why is this called "function" instead of "context"? "Context" seems to be clearer.
* Remove the space before the colon after variable names.
```suggestion
  var context: TraceContext? {
```

Reply:
+1, should this be context?

Reply:
Done.